File: Binary_Tree/543_diameter_of_tree/543.c
#include "../tree.h"
#include <stdlib.h>
#include <stdio.h>


int max(int a, int b) {
    if (a > b) {
        return a;
    }
    return b;
}

int recur(struct BinaryTree* root, int* maxDiameter) {
    if (root == NULL) {
        return 0;
    }
    int leftHeight = recur(root -> left, maxDiameter);
    int rightHeight = recur(root -> right, maxDiameter);
    *maxDiameter = max(*maxDiameter, leftHeight + rightHeight);
    return max(leftHeight, rightHeight) + 1;
}

int diameterOfTree(struct BinaryTree* root) {

    int * maxDiameter;
    int result = 0;
    maxDiameter = &result;
    recur(root, maxDiameter);
    return result;
}


int main() {
    int array[10] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    int arraySize = 10;
    int bufferArray[10] = {0};

    struct BinaryTree* root = build_tree(array, arraySize, bufferArray);
    print_tree(root, 1);
    
    int max_diameter = diameterOfTree(root);
    printf("\n diameter : %d \n", max_diameter);

}

File: Binary_Tree/543_diameter_of_tree/543.rs
mod tree;
use std::cmp::max;


fn max_depth(root: &Option<Box<tree::BinaryTree>>) -> usize {

    fn recur(root: &Option<Box<tree::BinaryTree>>, maxDiameter: &mut usize) -> usize {
        match root {
            None => return 0 as usize,
            Some(root) => {

                let leftHeight = recur(&root.left, maxDiameter);
                let rightHeight = recur(&root.right, maxDiameter);
                let max_diameter = max(*maxDiameter, leftHeight + rightHeight);
                *maxDiameter = max_diameter;
                return max(leftHeight, rightHeight) + 1;
            }
        }
    }
    let mut maxDiameter = 0;
    let _height = recur(root, &mut maxDiameter);
    return maxDiameter;
}


fn main() {
    let root = tree::build_tree(&mut vec!(0, 1, 2, 3, 4, 5, 6, 7, 8, 9));
    tree::print_tree(&root, 1);
    
    let diameter: usize = max_depth(&root);
    println!("diameter : {}", diameter);
}